[Object Spilling] Initial Iteration of S3 adapter.

rkooo567 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

This implements the initial (most primitive) prototype to support the S3 backend. It, unfortunately, is not properly working (extremely slow). There are 3 limitations in this PR, and I will have separate PRs to fix each of them. 

### This PR 
- builds a basic test framework. (Tests will run twice with different object spilling config)
- Make sure it "works" (though it has a really bad performance)
- Add some useful validation logic to ray.init path.

### Follow up
- It doesn't properly mock out S3 in tests. As a result, we cannot test it in our CI. There are several S3 mocking libraries (moto3 for example, https://github.com/spulec/moto). We should adopt this so that we can at least properly mock test S3 backends.
- It has "very poor" performance. It doesn't of course implement any fancy optimization like small objects fusion, but also doesn't implement the basic one like using thread pool or event queue. I tried with ThreadPoolExecutor and realized it wouldn't work because how this works is raylet pings the core worker cpp endpoint which subsequently invokes `restore_objects` methods. I will write a design doc before I implement this. 
- It doesn't modularize different adapters as well as defines proper interfaces. This also needs more discussion (for the right "interface").

## Related issue number

https://github.com/ray-project/ray/issues/9850

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

ericl commented on python/ray/external_storage.py:
```diff
@@ -46,6 +49,13 @@ def restore_spilled_objects(self, keys: List[bytes]):
         """
 
+    def validate(self):
+        """Check if the external storage configuration is valid.
```
Can we do this in the constructor instead?

rkooo567 replied:
Ah yeah. That's a good idea. Will fix.

ericl commented on python/ray/external_storage.py:
```diff
@@ -90,6 +103,119 @@ def restore_spilled_objects(self, keys):
+                             f"{self.directory_path}, doesn't exist.")
+
+
+class S3Storage(ExternalStorage):
```
Split into a separate file?

rkooo567 replied:
> It doesn't modularize different adapters as well as defines proper interfaces. This also needs more discussion (for the right "interface").

Will have a separate PR to refactor (and would like to get your sign off before I make that PR to avoid disagreement)

ericl replied:
Sounds good.

ericl commented on python/ray/external_storage.py:
```diff
@@ -90,6 +103,119 @@ def restore_spilled_objects(self, keys):
+                 get_config_override: dict = None,
+                 put_config_override: dict = None):
+        try:
+            import boto3
```
Have you considered using https://pypi.org/project/smart-open/ or https://pypi.org/project/s3fs/ instead of boto directly? They might implement a bunch of optimizations already.

rkooo567 replied:
I haven't. I will take a look at them.

ericl replied:
Thanks! Maybe we can try a simple test of upload performance of a large object (e.g., 100MB) to compare.

rkooo567 replied:
Sure! Btw, the biggest bottleneck now is not the s3 performance itself, but the fact that it is not parallelized (because each restore object is running on a main thread, and the whole thread is blocked until s3 ops are done. So the second IPC from object manager is blocked while we are waiting for the first one to be done). That should still be handled by a different pr, but I will see if I can utilize these libraries.

rkooo567 replied:
To parallelize, we should parallelize cpp -> cython -> python -> cython -> cpp -> reply path, which was not easy to be achieved by just using threadpool in python code

ericl replied:
Right-- I was thinking parallelism can be handled by having multiple IO workers. Also, if we coalesce many objects into big blocks, then the only thing that matters is large-chunk write performance.

rkooo567 replied:
Let me do some profiling with large enough object size (I assume 100MB?) first. I still feel like io worker level parallelization is not good enough (given how slow it is now). So, I will

1. Try smart open with sample test (benchmark) 
2. Profile s3 put/get
3. Write a One pager design doc to compare io worker vs io worker/thread pool vs io worker/event loop implementation.

rkooo567 replied:
Maybe 1 by this sprint and 2, 3 in the next

rkooo567 replied:
Ok. I read some doc/source code of both projects, and smart open seems to be more mature and has more options (one great part is it supports all major cloud providers). One drawback is that it doesn't really have asyncio compatible APIs (but if we only rely on Process-level parallelization or thread pool, it should be fine).

suquark replied:
boto3 actually supports parallelism by multipart uploading.

rkooo567 replied:
@suquark How big size objects are supposed to be to make it effective? Btw, parallelization I mentioned here is to context switching to the next download code while we are waiting for network IO.

ericl commented:
How about we go with smart_open for now then? Seems like it can be the
generic purpose fallback even if other options then out to be faster for s3
in the future.

On Wed, Oct 14, 2020, 11:22 PM SangBin Cho <notifications@github.com> wrote:

> *@rkooo567* commented on this pull request.
> ------------------------------
>
> In python/ray/external_storage.py
> <https://github.com/ray-project/ray/pull/11379#discussion_r505194857>:
>
> > +            the default boto3.resource('s3').Object(bucket_name, key).get()
> +        put_config_override(dict): Configuration dict that will override
> +            the default boto3.resource('s3').Object(bucket_name, key).put()
> +    Raises:
> +        RayError if it fails to setup a S3 client. For example, if boto3 is
> +        not downloaded, it raises an RayError. It can also raise S3 related
> +        exceptions.
> +    """
> +
> +    def __init__(self,
> +                 bucket_name: str,
> +                 prefix: str = "ray_spilled_object_",
> +                 get_config_override: dict = None,
> +                 put_config_override: dict = None):
> +        try:
> +            import boto3
>
> Ok. I read some doc/source code of both projects, and smart open seems to
> be more mature and has more options (one great part is it supports all
> major cloud providers). One drawback is that it doesn't really have asyncio
> compatible APIs (but if we only rely on Process-level parallelization or
> thread pool, it should be fine).
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ray-project/ray/pull/11379#discussion_r505194857>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAADUSUYLIIUEJVKXTR6WFLSK2ILJANCNFSM4SP2RV2A>
> .
>

rkooo567 commented:
Finished the smart_open implementation. Here's the result. S3 is almost 600 times slower for `get`, and 200 times slower for `put` with multi-part download for 96MB objects. 

fs
```
Object spilling benchmark for the config {'type': 'filesystem', 'params': {'directory_path': '/tmp'}}
Spilling 50 number of objects of size 12582912B takes 4.317760701999999 seconds with 10 number of io workers.
Getting all objects takes 2.3372979060000008 seconds.
```

Smart open
```
Object spilling benchmark for the config {'type': 's3', 'params': {'bucket_name': 'sang-object-spilling-test'}}
Spilling 50 number of objects of size 12582912B takes 743.244563962 seconds with 10 number of io workers.
Getting all objects takes 1165.6987561679998 seconds.
```

Some possible root causes;
- https://github.com/ray-project/ray/issues/11495; It looks like there are only 2 IO workers although I specified 10 max workers.
- The object upload happened on my laptop. It will be much faster if I do it in EC2 in the same region & VPN as S3.

We need more performance analysis, but I won't handle them in this PR.

That says, after this PR I will
1. Fix the IO worker's issue.
2. Benchmark within an EC2 instance with different object sizes.
3. Implement object fusion.
4. Do more performance analysis and optimization.

ericl commented on python/ray/external_storage.py:
```diff
@@ -90,6 +129,223 @@ def restore_spilled_objects(self, keys):
+                self._put_object_to_store(metadata, buf_len, file_like, ref)
+
+
+class S3StorageSmartOpenImpl(ExternalStorage):
```
```suggestion
class ExternalStorageSmartOpenImpl(ExternalStorage):
```

smart_open works with many storage systems; so we should not include "S3" in the name here.

rkooo567 replied:
Ah, I see. So you suggest making this implementation completely independent from storages. That sounds like a good idea. I will address all the comments below.

ericl commented on python/ray/external_storage.py:
```diff
@@ -90,6 +129,223 @@ def restore_spilled_objects(self, keys):
+        return {}
+
+
+class S3StorageBotoCoreImpl(ExternalStorage):
```
Yeah, let's remove this for now.

ericl commented on python/ray/external_storage.py:
```diff
@@ -90,6 +129,223 @@ def restore_spilled_objects(self, keys):
+                 override_transport_params: dict = None,
+                 track_tail_latency: bool = False):
+        try:
+            import boto3
```
Remove any reference to boto or s3.

ericl commented on python/ray/external_storage.py:
```diff
@@ -90,6 +129,223 @@ def restore_spilled_objects(self, keys):
+
+    def get_stats(self):
+        # TODO(sang): Implement it for benchmark.
+        return {}
```
Let's remove stats tracking for now; it is fine to track in the benchmark script.

ericl commented on python/ray/external_storage.py:
```diff
@@ -90,6 +129,223 @@ def restore_spilled_objects(self, keys):
+            self._put_object_to_store(metadata, buf_len, file_like, ref)
+
+    def _build_s3_url(self, key):
+        return f"s3://{self.bucket_name}/{key}"
```
Instead of bucket name and key, the user can provide a storage prefix, e.g. "s3://bucket/foo/"

ericl commented on python/ray/external_storage.py:
```diff
@@ -101,12 +357,21 @@ def setup_external_storage(config):
         storage_type = config["type"]
         if storage_type == "filesystem":
             _external_storage = FileSystemStorage(**config["params"])
+        elif storage_type == "s3":
```
smart_open?

rkooo567 commented:
Addressed all review. The next step;

- Finish writing a design for fusion small objects.
- Figure out why only 2 io workers are used although I specified a high number.
- Actual performance benchmark within Amazon VPN that is within S3 region. (and tuning)

ericl commented on python/ray/external_storage.py:
```diff
@@ -56,13 +77,25 @@ def spill_objects(self, object_refs):
     def __init__(self, directory_path):
         self.directory_path = directory_path
         self.prefix = "ray_spilled_object_"
+        if not os.path.exists(self.directory_path):
```
We can probably just raise ValueError for now

ericl commented on python/ray/external_storage.py:
```diff
@@ -91,6 +124,84 @@ def restore_spilled_objects(self, keys):
+            raise ExternalStorageConfigInvalidError(
+                "S3 storage is chosen to be a object spilling "
+                "external storage, but smart_open "
+                "is not downloaded.")
```
It would be better to not catch this error and just allow it to propagate. Otherwise, it is unclear what the actual cause is.

ericl reviewed (approved):
Looks good but I think it would be good to simplify the error handling.
